src/backend/fs.rs

```diff
@@ -3,151 +3,127 @@
 use std::{
     fs::{self, OpenOptions},
     io::{self, Read, Write},
-    //path::Path,
     path::{Path, PathBuf},
 };
 
-use anyhow::{anyhow, Result};
-use carbonado::{
-    constants::{Format, FEC_M},
-    fs::Header,
-    structs::Encoded,
-};
+use anyhow::{anyhow, Error, Result};
+use axum::body::Bytes;
+use carbonado::{constants::Format, file::Header, structs::Encoded};
+use futures_util::{stream::BoxStream, StreamExt, TryStreamExt};
 use log::{debug, trace};
 use rayon::prelude::*;
 use secp256k1::{ecdh::SharedSecret, PublicKey, SecretKey};
 
 use crate::{config::SYS_CFG, prelude::*};
 
-pub async fn write_file(pk: &Secp256k1PubKey, file_bytes: &[u8]) -> Result<Blake3Hash> {
-    debug!("Write file, len: {}", file_bytes.len());
+pub type FileStream<'a> = BoxStream<'a, Result<Bytes>>;
 
-    trace!("Create a shared secret using ECDH");
+pub async fn write_file<'a>(
+    pk: &Secp256k1PubKey,
+    file_stream: FileStream<'a>,
+) -> Result<Blake3Hash> {
+    trace!("write_file, create a shared secret using ECDH");
     let sk = SYS_CFG.private_key;
-    let ss = SharedSecret::new(&pk.into_inner(), &sk);
-    let write_pk = PublicKey::from_secret_key_global(&SecretKey::from_slice(&ss.secret_bytes())?);
+    let ss = SharedSecret::new(&pk.into_inner(), &sk).secret_bytes();
+    let write_pk = PublicKey::from_secret_key_global(&SecretKey::from_slice(&ss)?);
 
     let pk_bytes = write_pk.serialize();
     let (x_only_pk, _) = write_pk.x_only_public_key();
 
-    let file_hash = Blake3Hash(blake3::keyed_hash(&x_only_pk.serialize(), file_bytes));
-
-    trace!("TODO: Check if file catalog already exists");
-    // let path = SYS_CFG
-    //     .volumes
-    //     .get(0)
-    //     .expect("First volume present")
-    //     .path
-    //     .join(CATALOG_DIR)
-    //     .join(file_hash.to_string());
+    trace!("Initialize Blake3 keyed hasher");
+    let mut file_hasher = blake3::Hasher::new_keyed(&x_only_pk.serialize());
```

Comment:
Hasher simply updates the hash incrementally for each chunk?

Reply:
Correct